bluemira/optimisation/_algorithm.py

```diff
@@ -0,0 +1,73 @@
+# bluemira is an integrated inter-disciplinary design tool for future fusion
+# reactors. It incorporates several modules, some of which rely on other
+# codes, to carry out a range of typical conceptual fusion reactor design
+# activities.
+#
+# Copyright (C) 2021-2023 M. Coleman, J. Cook, F. Franza, I.A. Maione, S. McIntosh,
+#                         J. Morris, D. Short
+#
+# bluemira is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# bluemira is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with bluemira; if not, see <https://www.gnu.org/licenses/>.
+"""
+Enumeration of supported optimiser algorithms.
+
+Note that it may not be the case that all backends support all
+algorithms. Each `Optimiser` implementation must state which of these
+algorithms it supports.
+"""
+
+from __future__ import annotations
+
+import enum
+
+
+class Algorithm(enum.Enum):
+    """Enumeration of available optimisation algorithms."""
+
+    SLSQP = enum.auto()
+    COBYLA = enum.auto()
+    SBPLX = enum.auto()
+    MMA = enum.auto()
+    BFGS = enum.auto()
+    DIRECT = enum.auto()
+    DIRECT_L = enum.auto()
+    CRS = enum.auto()
+    ISRES = enum.auto()
+
+    @classmethod
+    def from_string(cls, s: str) -> Algorithm:
+        """
+        Get enumeration value from the given string.
+
+        Parameters
+        ----------
+        s: str
+            String representation of the enum
+
+        Returns
+        -------
+        e: Algorithm
+            The enumeration.
+
+        Raises
+        ------
+        ValueError
+            If the string does not correspond to an enum name.
+        """
+        try:
+            return getattr(cls, s)
+        except AttributeError:
+            if s == "DIRECT-L":
+                # special case for backward compatibility
+                return Algorithm.DIRECT_L
+            raise ValueError(f"No such Algorithm value '{s}'.")
```

Comment:
I dont know whether it might be better to modify the `__getitem__` in a metaclass because currently this is possible
Algorithm["DIRECT_L"] but Algorithm["DIRECT-L"] would fail

```python
class AlgorithmMeta(enum.EnumMeta):
    def __getitem__(self, s: str) :
        try:
            return super().__getitem__(s)
        except AttributeError:
            if s == "DIRECT-L":
                # special case for backward compatibility
               return self.DIRECT_L
            raise ValueError(f"No such Algorithm value '{s}'.")

class Algorithm(enum.Enum, metaclass=AlgorithmMeta):
```